game/scripts/vscripts/abilities/oaa_arcane_orb.lua

```diff
@@ -31,175 +43,227 @@ function modifier_oaa_arcane_orb:RemoveOnDeath()
   return false
 end
 
-function modifier_oaa_arcane_orb:OnCreated()
-  if IsServer() then
-    if not self.procRecords then
-      self.procRecords = {}
-    end
-    self.parentOriginalProjectile = self:GetParent():GetRangedProjectileName()
-    Debug.EnabledModules["abilities:oaa_arcane_orb"] = false
-  end
-end
-
-modifier_oaa_arcane_orb.OnRefresh = modifier_oaa_arcane_orb.OnCreated
-
 function modifier_oaa_arcane_orb:DeclareFunctions()
   return {
     MODIFIER_EVENT_ON_ATTACK_START,
     MODIFIER_EVENT_ON_ATTACK,
     MODIFIER_EVENT_ON_ATTACK_LANDED,
-    MODIFIER_EVENT_ON_ATTACK_FAIL
+    MODIFIER_EVENT_ON_ATTACK_FINISHED
   }
 end
 
-function modifier_oaa_arcane_orb:OnAttackStart(keys)
+function modifier_oaa_arcane_orb:OnAttack(event)
   local parent = self:GetParent()
+  local ability = self:GetAbility()
 
-  if keys.attacker ~= parent then
+  if event.attacker ~= parent then
     return
   end
 
-  local ability = self:GetAbility()
-  local target = keys.target
-  -- Wrap in function to defer evaluation
-  local function autocast()
-    return (
-      target.GetUnitName and -- Check for existence of GetUnitName method to determine if target is a unit
-      ability:GetAutoCastState() and
-      not parent:IsSilenced() and
-      ability:IsOwnersManaEnough() and
-      ability:IsOwnersGoldEnough(parent:GetPlayerOwnerID()) and
-      ability:IsCooldownReady() and
-      ability:CastFilterResultTarget(target) == UF_SUCCESS
-    )
+  if parent:GetCurrentActiveAbility() ~= ability then
+    return
+  end
+
+  -- To prevent crashes:
+  local target
+  if event.target == nil then
+    return
+  else
+    target = event.target
   end
 
-  if parent:GetCurrentActiveAbility() ~= ability and not autocast() then
+  if target:IsNull() then
     return
   end
 
-  -- Add modifier to change attack sound
-  parent:AddNewModifier(parent, ability, "modifier_oaa_arcane_orb_sound", {})
-  -- Set projectile
-  parent:ChangeAttackProjectile()
+  -- Check for existence of GetUnitName method to determine if target is a unit
```

Comment:
Maybe clarify this with "(and not, e.g., an item)

Reply:
I added more comments to clarify. I wonder what other **attackable** entities beside items don't have a GetUnitName method.